### PR TITLE
Catch eligibility and change interface

### PR DESCRIPTION
**Pull request by Francois-Gomis**

### CODE REVIEW

File: Helpers/QuoteHelper.php [hunk 1]
```diff
@@ -52,6 +54,11 @@ class QuoteHelper
      */
     private $quoteRepository;
 
+    /**
+     * @var null
+     */
+    private $quoteId;
```
Comment (syjust) [inline Helpers/QuoteHelper.php:60]: You are replacing `$quote` by `$quoteId` 😛

File: Helpers/QuoteHelper.php [hunk 2]
```diff
@@ -52,6 +54,11 @@ class QuoteHelper
      */
     private $quoteRepository;
 
+    /**
+     * @var null
```
Comment (syjust) [inline Helpers/QuoteHelper.php:58]: please add a type `null|int` ?

File: Helpers/QuoteHelper.php [hunk 3]
```diff
@@ -93,30 +101,40 @@ public function getQuote($cartId=null):?\Magento\Quote\Model\Quote
...
+    public function setEligibilityQuoteId($quoteId):void
+    {
+        $this->quoteId = $quoteId;
     }
 
+
     /**
      * Get quote from session if is define
-     * @return \Magento\Quote\Model\Quote|null
+     * @return \Magento\Quote\Api\Data\CartInterface|null
```
Comment (syjust) [inline Helpers/QuoteHelper.php:119]: you can use `CartInterface|null` instead FQDN here

File: Helpers/QuoteHelper.php [hunk 4]
```diff
@@ -128,15 +146,14 @@ public function getQuoteByContextUserId():?\Magento\Quote\Model\Quote
 
     /**
      * Load quote with cartID
-     * @param $cartId
-     * @return \Magento\Quote\Model\Quote\Interceptor|null
+     * @return \Magento\Quote\Api\Data\CartInterface|null
```
Comment (syjust) [inline Helpers/QuoteHelper.php:149]: you can use `CartInterface|null` instead FQDN here

File: Helpers/Eligibility.php [hunk 5]
```diff
@@ -134,10 +121,16 @@ public function __construct(
...
      * @throws RequestError
-     * @throws InvalidArgumentException
+     * @throws \InvalidArgumentException
      */
     private function getPlansEligibility(): array
     {
+        try {
+            $quote = $this->getEligibilityQuote();
+        } catch (\InvalidArgumentException $e) {
+            throw new \InvalidArgumentException($e->getMessage());
```
Comment (syjust) [inline Helpers/Eligibility.php:131]: don't use FQDN

add import with `\InvalidArgumentException`

File: Helpers/Eligibility.php [hunk 6]
```diff
@@ -216,9 +209,11 @@ private function getPlansEligibility(): array
      */
     public function checkEligibility(): bool
     {
-
-        if(!isset($this->quote)){
-            throw new \InvalidArgumentException('No Quote for eligibility');
+        $this->logger->info('checkEligibility',[]);
+        try {
+            $quote = $this->getEligibilityQuote();
+        } catch (\InvalidArgumentException $e) {
```
Comment (syjust) [inline Helpers/Eligibility.php:215]: don't use FQDN

add import with `\InvalidArgumentException`

File: Helpers/Eligibility.php [hunk 7]
```diff
@@ -534,39 +534,18 @@ private function buildType($installmentCount,$isDeferred):string
...
      */
-    private function getEligibilityQuote()
+    private function getEligibilityQuote():?CartInterface
     {
-        if(isset($this->quote)){
-            return $this->quote;
+        $quote=$this->quoteHelper->getQuote();
+        if(!isset($quote)){
+            $this->logger->info('No Quote for eligibility',[]);
+            throw new \InvalidArgumentException('No Quote for eligibility');
```
Comment (syjust) [inline Helpers/Eligibility.php:546]: don't use FQDN

add import with `\InvalidArgumentException`

File: Helpers/QuoteHelper.php [hunk 8]
```diff
@@ -93,30 +101,40 @@ public function getQuote($cartId=null):?\Magento\Quote\Model\Quote
...
+        if($sessionQuoteId != null){
             $quote = $this->checkoutSession->getQuote();
         }
         return $quote;
     }
 
 
     /**
-     * @return \Magento\Quote\Model\Quote|null
+     * @return \Magento\Quote\Api\Data\CartInterface|null
```
Comment (syjust) [inline Helpers/QuoteHelper.php:134]: you can use `CartInterface|null` instead FQDN here

File: composer.json [hunk 9]
```diff
@@ -6,7 +6,7 @@
         "alma/alma-php-client": ">=1.3.1"
     },
     "type": "magento2-module",
-    "version": "2.2.0",
+    "version": "2.2.1",
```
Comment (syjust) [inline composer.json:9]: you forgot to update `php` min version to `>=7.1`

File: Helpers/QuoteHelper.php [hunk 10]
```diff
@@ -93,30 +101,40 @@ public function getQuote($cartId=null):?\Magento\Quote\Model\Quote
...
      * @throws \Magento\Framework\Exception\NoSuchEntityException
```
Comment (syjust) [inline Helpers/QuoteHelper.php:135]: don't use FQDN

add import with `\Magento\Framework\Exception\NoSuchEntityException`

File: Helpers/QuoteHelper.php (hunk 1 above)
Comment (Francois-Gomis) [reply Helpers/QuoteHelper.php:60]: I have remote quote from eligibility  - I have save quoteId in quoteHelper

File: Helpers/Eligibility.php (hunk 7 above)
Comment (Francois-Gomis) [reply Helpers/Eligibility.php:546]: ?

File: Helpers/QuoteHelper.php [hunk 11]
```diff
@@ -128,15 +146,14 @@ public function getQuoteByContextUserId():?\Magento\Quote\Model\Quote
 
     /**
      * Load quote with cartID
-     * @param $cartId
-     * @return \Magento\Quote\Model\Quote\Interceptor|null
+     * @return \Magento\Quote\Api\Data\CartInterface|null
      * @throws \Magento\Framework\Exception\NoSuchEntityException
```
[the remark this answers is not in the document]
Comment (syjust) [reply Helpers/QuoteHelper.php:150]: there is a `try catch` that catch all `Exception`

your method will never throws `NoSuchEntityException`

you can remove it from the docblock

File: Helpers/Eligibility.php (hunk 5 above)
Comment (syjust) [reply Helpers/Eligibility.php:131]: you should use `use \InvalidArgumentException;` in the header of your class file.

then, you can after call the `InvalidArgumentException` without FQDN (without `\` suffix)

**Review by syjust — COMMENTED** [review]
you should use use `\InvalidArgumentException;` in the header of your class file.

then, you can after call the `InvalidArgumentException` without FQDN (without \ suffix)


in `getQuoteById()` method

there is a `try catch` that catch all `Exception`s

your method will never throws `NoSuchEntityException`

you can remove it from the docblock

**Pull request merged**